### ISSUE TITLE
Tracker: Support free-threaded CPython builds

### ISSUE THREAD

**Issue by andfoy** [scipy/scipy#20669]
This issue aims to collect all the issues and PRs related for the upcoming free-threaded CPython 3.13 (a.k.a. "no-GIL") builds. 

The outline of this issue will detail the compatibility of each module in the SciPy namespace against nogil builds, specially with class APIs, which may suffer of concurrency issues due to state sharing when multiple threads share the same object instances.

Until now, as of https://github.com/scipy/scipy/pull/20611 all the test suite is passing using the latest NumPy after https://github.com/numpy/numpy/pull/26348. 

Resources:

* [PEP 703](https://peps.python.org/pep-0703/)
* https://py-free-threading.github.io/ (site with extensive guidance on how to approach support for free-threaded CPython
* [Fork of numpy 1.24 with early nogil support](https://github.com/numpy/numpy/compare/v1.24.0...colesbury:numpy:v1.24.0-nogil)
* Corresponding tracker in NumPy: https://github.com/numpy/numpy/issues/26157

## CI jobs & wheels

- [x] regular CI job for free-threaded CPython on Linux: gh-20822
- [x] wheel builds for:
    - [x] Linux (`manylinux` and `musllinux`): gh-20882
    - [x] macOS (4x: `x86-64` and `arm64`, with OpenBLAS and Accelerate): gh-20951
    - [x] Windows (1x, 64-bit only since we don't support 32-bit): gh-21793
- [x] Enable [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel) in a CI job after the whole test suite cleanly runs under it. See the `parallel: 1` entry under the `free-threaded` job in `linux.yml` (introduced in [gh-21496](gh-21496)). In progress in gh-22658
- [x] gh-22758

## Code changes

- [x] Marking extension modules as supporting running without the GIL:
    - [x] C/C++/Cython: gh-21199 (a few loose ends taken care of in [gh-21961](https://github.com/scipy/scipy/pull/21961))
    - [x] `f2py`-generated: gh-21943
    - [x] `pythran`-generated: (blocked on support in `pythran`, xref [pythran#2258](https://github.com/serge-sans-paille/pythran/issues/2258))
- [x] Remove `thread_unsafe `markers with a `cython#6506` comment after the fix for [cython#6506](https://github.com/cython/cython/issues/6506) is merged and a new Cython nightly and conda-forge package are available. Done in gh-22399.

Per-submodule changes/support:

- [ ] `_lib`
    - [x] Issues uncovered by `pytest-run-parallel` were fixed in  [gh-21496](https://github.com/scipy/scipy/pull/21496)
    - [x] gh-21218
    - [ ] File-based `MessageStream` implementation looks shaky under `pytest-repeat` 
    - [ ] `ccallback.h` contains thread-local storage; code that uses it should gain multi-threaded tests
    - [x] gh-22547
- [x] `cluster` (only Cython code, audited, no problems found); clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496)
- [x] `constants` (pure Python)
- [x] `datasets` (pure Python); clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496)
- [x] `differentiate` (pure Python); clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496)- [x] `fft` (audited, and has a decent amount of multi-threaded tests); clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496)
- [x] `fftpack` (audited, only has one Cython file, and the cache in it no longer exists); clean under `pytest-run-parallel` after [gh-21496](https://github.com/scipy/scipy/pull/21496))
- [x] `integrate` (a lot of thread-safety issues fixed in [gh-21496](https://github.com/scipy/scipy/pull/21496)
- [x] `interpolate`: 
    - [x] gh-20671
    - [x] clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496)
- [x] `io` (clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496))
- [x] `linalg`
    - [x] (clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496))
    - [x] Verify safety of low level `linalg.blas` and `linalg.lapack` functions: gh-21937
    - [x] gh-21936 
    - [x] gh-21479
- [x] `misc` (pure Python, clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496))
- [x] `ndimage` (clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496))
- [x] `odr` (clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496))
- [x] `optimize` (clean under `pytest-run-parallel` after [gh-21496](https://github.com/scipy/scipy/pull/21496))
- [x] `signal` (clean under `pytest-run-parallel` after  [gh-21496](https://github.com/scipy/scipy/pull/21496))
    - [x] `_sigtools`: gh-21160
- [x] `sparse` (clean under `pytest-run-parallel` after [gh-21954](https://github.com/scipy/scipy/pull/21954))
- [x] `spatial` (clean under `pytest-run-parallel` after [gh-21955](https://github.com/scipy/scipy/pull/21955))
    - [x] `_Qhull` → gh-20619, gh-20611
    - [x] `KDTree/cKDTree` → gh-24799
- [x] `special`
    - [x] Test suite made clean (except for error handling) under `pytest-run-parallel` after [gh-21956](https://github.com/scipy/scipy/pull/21956))
    - [x] Make error handling thread-safe (done in [gh-22742](https://github.com/scipy/scipy/pull/22742))
- [x] `stats`
    - [x] tests clean under `pytest-run-parallel`: gh-22125
    - [x] thread-safety fixes: gh-22175

Note that the approach is to scan for known issues, audit code that is likely to be problematic, add some threaded tests (lightly; will be far from comprehensive), and then test at a higher level (e.g., also rely on scikit-learn, scikit-image, etc. exercising SciPy functionality). We may go back through and add new test approaches as they become available and better understood (e.g., use of ThreadSanitizer). _EDIT: we can now use [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel) for detecting a number of types of thread-safety problems automatically._

## Documentation

- [x] Add a docs section on the overall state of thread safety with and without free-threading in SciPy: gh-21941
- [x] For functionality that we know is unsafe, document this in the API reference docs for the functionality in question. (made a start in [gh-21941](https://github.com/scipy/scipy/pull/21941))

**Comment by andfoy** [comment]
Here is a list of classes in SciPy, which can be of interest, since they could have shared state that can be mutated under multithreading: https://gist.github.com/andfoy/955bf6ceee8437b2092df19202490981. Feel free to comment if you have any hints or suspicions that a class might be non-thread safe

I extracted them by means of the following introspection script:

<details>
<summary>Script</summary>

```python

import scipy
from inspect import isclass, ismodule, getmembers

cls_set = set({})
visited = set({})
queue = [scipy]

while queue != []:
    mod_or_class = queue.pop(0)
    if mod_or_class in visited:
        continue
    print(mod_or_class)
    if ismodule(mod_or_class) and mod_or_class.__name__.startswith('scipy'):
        for _, mem in getmembers(mod_or_class):
            if ismodule(mem) and mem.__name__.startswith('scipy'):
                queue.append(mem)
            elif isclass(mem):
                queue.append(mem)
    elif isclass(mod_or_class):
        cls_module = mod_or_class.__module__
        cls_name = mod_or_class.__name__
        fully_qual = f'{cls_module}.{cls_name}'
        if cls_module.startswith('scipy') and fully_qual not in cls_set:
            print(fully_qual)
            cls_set |= {fully_qual}
    visited |= {mod_or_class}

classes = sorted(cls_set)
for cls_name in classes:
      print(cls_name)
```
</details>

**Comment by rgommers** [comment]
> Feel free to comment if you have any hints or suspicions that a class might be non-thread safe

A couple of thoughts:
- Distributions in `scipy.stats` store a random number generator as internal state
- Fortran code may contain `COMMON` blocks, and perhaps other ways of storing global/shared state. My expectation is that a lot of our Fortran code isn't safe It's be great if a Fortran expert could comment on that (@ev-br or @ilayn?)
- We don't have many caches, but those we have require checking. E.g. the `optimize.shgo` implementation has one.
- Uarray should be using thread-local state correctly already IIRC, and isn't used a lot. So no need to prioritize that one at least.

**Comment by ev-br** [comment]
`COMMON` blocks are relatively rare: the only non-legacy submodule which does not have a more modern replacement is ODRPACK (`integrate/dop` and `odepack` have an alternative in `solve_ivp; `mvndst` is likely ready or almost ready to be removed according to https://github.com/scipy/scipy/issues/18566 )


```
$ grep "COMMON /" . -rn
./scipy/integrate/mach/d1mach.f:17:      COMMON /D9MACH/ CRAY1
./scipy/integrate/dop/dopri5.f:372:      COMMON /CONDO5/XOLD,HOUT
./scipy/integrate/dop/dopri5.f:629:      COMMON /CONDO5/XOLD,H
./scipy/integrate/dop/dop853.f:546:      COMMON /CONDO8/XOLD,HOUT
./scipy/integrate/dop/dop853.f:862:      COMMON /CONDO8/XOLD,H
./scipy/integrate/odepack/zvode.f:978:C     COMMON /ZVOD01/ RVOD(50), IVOD(33)
./scipy/integrate/odepack/zvode.f:979:C     COMMON /ZVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/zvode.f:1182:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:1190:      COMMON /ZVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/zvode.f:1902:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:1910:      COMMON /ZVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/zvode.f:2063:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:2071:      COMMON /ZVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/zvode.f:2470:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:2642:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:2854:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:2862:      COMMON /ZVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/zvode.f:3080:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:3088:      COMMON /ZVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/zvode.f:3342:      COMMON /ZVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13), ETA,
./scipy/integrate/odepack/zvode.f:3415:      COMMON /ZVOD01/ RVOD1(50), IVOD1(33)
./scipy/integrate/odepack/zvode.f:3416:      COMMON /ZVOD02/ RVOD2(1), IVOD2(8)
./scipy/integrate/odepack/vode.f:973:C     COMMON /DVOD01/ RVOD(48), IVOD(33)
./scipy/integrate/odepack/vode.f:974:C     COMMON /DVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/vode.f:1168:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:1176:      COMMON /DVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/vode.f:1882:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:1890:      COMMON /DVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/vode.f:2042:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:2050:      COMMON /DVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/vode.f:2449:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:2621:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:2832:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:2840:      COMMON /DVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/vode.f:3061:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:3069:      COMMON /DVOD02/ HU, NCFN, NETF, NFE, NJE, NLU, NNI, NQU, NST
./scipy/integrate/odepack/vode.f:3329:      COMMON /DVOD01/ ACNRM, CCMXJ, CONP, CRATE, DRC, EL(13),
./scipy/integrate/odepack/vode.f:3403:      COMMON /DVOD01/ RVOD1(48), IVOD1(33)
./scipy/integrate/odepack/vode.f:3404:      COMMON /DVOD02/ RVOD2(1), IVOD2(8)
./scipy/odr/odrpack/d_test.f:28:      COMMON /TSTSET/ NTEST
./scipy/odr/odrpack/d_test.f:151:      COMMON /SETID/SETNO
./scipy/odr/odrpack/d_test.f:152:      COMMON /TSTSET/ NTEST
./scipy/odr/odrpack/d_test.f:974:      COMMON /SETID/SETNO
./scipy/odr/odrpack/d_test.f:1843:      COMMON /SETID/SETNO
./scipy/stats/mvndst.f:203:      COMMON /DKBLCK/IVLS
```

"implicit save" might be a build problem though, not sure TBH: https://github.com/scipy/scipy/issues/6882#issuecomment-269102164

**Comment by ilayn** [comment]
About `mvndst.f` ; I was about to translate it but I can't remember where I am told that it is already implemented in somewhere else and marked that issue might be Warren or Matt. We also have #8367 about this. But I really don't know what the status is about it. I can squeeze that in and complete the `stats` translation suite for quick dopamine. 

`COMMON`, `IMPLICIT SAVE` and `SAVE` are really annoying "features" of old code. But `COMMON` is the lesser evil. The others are just pure pain. So I would actually like some guidance about prioritizing this Fortran saga as we are approaching to array based parts of the translation. My unfortunately reallife/work-affected schedule is 

1. `id_dist` (hoping to finish this asap for 1.14)
2. `slsqp` (this is already translated but I'm blocked on the python side as I have not found a way to disentangle the `_minimize.py` yet; it is really a lot of historical crust about constraints and jacobian and classes etc. Problem is mostly my IQ)
3. propack (real part done, complex is waiting)
4. arpack (I don't expect too much difficulty in this one since it is a "simpler" version of propack)
5. and then no idea what to start, I hope we can get rid of ODRPACK by reading F90 version. integrate is your playground so not sure what the situation is in that module.

**Comment by rgommers** [comment]
@ilayn sorry for the slow reply. I think ARPACK/PROPACK are probably the biggest bang for the buck. They are full of bugs that are impossible to fix. It will help with introducing support for free-threaded CPython, and at the same time (more important even) it will fix a host of crashes and failures. There's also several unfinished PRs open that attempt to upgrade ARPACK from 3.3.0 to 3.7.0/3.8.0 that ran into trouble.

**Comment by ilayn** [comment]
Yes, I finally got the distance on `id-dist` needed. This week I am hoping to finish. Then ARPACK/PROPACK it is.

**Comment by albanD** [comment]
Running with pydebug build as of two weeks ago and using ua_function, 1.15.0dev0 crashes on exit with
```
(null):0: _Py_NegativeRefcount: Assertion failed: object has negative ref count
<object at 0x2000e12e2b0 is freed>
Fatal Python error: _PyObject_AssertFailed: _PyObject_AssertFailed
Python runtime state: finalizing (tstate=0x00000000009bab80)
```

From @colesbury comment that 3.13t sets all module as being immortal, the explanation is that https://github.com/scipy/scipy/blob/941d5b08614841a213019990b6ceee83a05d6dcc/scipy/_lib/_uarray/_uarray_dispatch.cxx#L1774 never gets called and so the static string from https://github.com/scipy/scipy/blob/941d5b08614841a213019990b6ceee83a05d6dcc/scipy/_lib/_uarray/_uarray_dispatch.cxx#L145 outlives the python runtime while still having a strong reference.
When we get to static variable de-alloc, we end up calling the py_ref destructor https://github.com/scipy/scipy/blob/941d5b08614841a213019990b6ceee83a05d6dcc/scipy/_lib/_uarray/_uarray_dispatch.cxx#L38 (without taking the gil btw) while the python runtime is already dead (and in this case the string already properly de-allocated by the runtime) and thus we detect it to have negative refcount.
Since the m_free not being called is not going to be fixed, I guess the right fix is to have the static variable have an owning reference to the py_obj but explicitly leak it (never call decref on it).

This would be the same treatment being applied to PyTorch's static variables in https://github.com/pytorch/pytorch/pull/130341 and other places like https://github.com/pytorch/pytorch/blob/0b134c15cdf8cb4c2c7ffed1c2b6518de44382bf/torch/csrc/autograd/python_hook.cpp#L105

**Comment by rgommers** [comment]
Thanks a lot for the detailed report and suggested way of fixing it @albanD! I've moved this to a separate issue: gh-21214

**Comment by hawkinsp** [comment]
I note 1.15 shipped Linux x86_64 cp313t wheels but no Linux aarch64. Was this an oversight?

**Comment by rgommers** [comment]
@hawkinsp not quite an oversight, more a decision to delay doing a bunch of extra work for a more niche platform at this stage. Linux aarch64 wheels are built separately from all the other wheels on Cirrus CI, rather than on GitHub. So it's separate workflows to maintain, and also paid infra where we currently squeeze wheel builds just under a 60 minute limit.

We can revisit though if it's actually a problem for you, and possibly ship those wheels for 1.15.2

**Comment by h-vetinari** [comment]
Not the same installer, but I'll note that conda-forge has 3.13t builds for all our platforms, including linux-aarch64

**Comment by hawkinsp** [comment]
Well, it's not a hugely urgent problem, but it is a problem. We're trying to set up CI builds of our own wheels under cp313t; Linux aarch64 is a platform we support and of growing popularity, and scipy is one of our deps. So it'd be nice to get this done sometime!

**Comment by andyfaff** [comment]
There's a wider question here w.r.t number of wheels that are built for each release. (I think we have had this conversation a little bit before).
Making free threaded wheels eventually doubles the matrix size. There are consequences for storage, build resources, etc. I'm not saying that they shouldn't be built, rather we should be aware of the cost when we do any expansion of the matrix

**Comment by rgommers** [comment]
> Making free threaded wheels eventually doubles the matrix size.

That won't happen any time soon - and it may never happen. The free-threaded build is planned to become the default build of CPython, and that avoids the double builds. Avoiding doubling the number of wheels was one of the main concerns when PEP 703 was discussed and decided on.

> rather we should be aware of the cost when we do any expansion of the matrix

Yes, avoiding the extra cost and work was the reason we delayed this. I think we should switch to GitHub Linux arm64 runners first ([now available in public preview](rather we should be aware of the cost when we do any expansion of the matrix)) and then add the `cp313t` variants. WDYT @andyfaff? I'm happy to do the work, unless you want to.

**Comment by rgommers** [comment]
It was fairly easy to do, so I gave it a try in https://github.com/rgommers/scipy/tree/wheels-gha-aarch64. The wheel builds work and are about 25% faster than on the x86-64 runners. There is some flakiness in the runners still though, I saw 2 successful builds and one run where docker choked, similar to what other people reported seeing in https://github.com/pypa/cibuildwheel/issues/2132.

**Comment by lesteve** [comment]
FWIW, we have switched to github arm runners for wheels in scikit-learn https://github.com/scikit-learn/scikit-learn/pull/30658 and so far it seems to be working fine for us. In particular I haven't noticed the flakiness you mentioned but maybe I have missed it.

**Comment by rgommers** [comment]
Thanks @lesteve, that is useful context. I've just seen it once (with 3 successful runs now), so it may not be a frequent issue. The occasional "re-run job" button push isn't the end of the world perhaps.

**Comment by andyfaff** [comment]
I won't be able to do any deep dive writing for another couple of weeks for any CI dev work. Adding aarch64 should be quite simple for a CI run and for a wheel build. However free threading variant does require a bit of massaging.

**Comment by rgommers** [comment]
No worries, thanks @andyfaff. I'll try to take care of it within the next few days.

**Comment by tylerjereddy** [comment]
> The occasional "re-run job" button push isn't the end of the world perhaps.

Yeah, needed in releases often enough, especially given the growing size of our matrix and the likelihood of at least one flake somewhere.

**Comment by rgommers** [comment]
> We can revisit though if it's actually a problem for you, and possibly ship those wheels for 1.15.2

@hawkinsp there are `cp313t` aarch64 manylinux/musllinux nightly wheels up at https://anaconda.org/scientific-python-nightly-wheels/scipy/ now, and they'll land in 1.15.2 when that's released (no exact timing, a few week to 1-2 months typically).

**Comment by hawkinsp** [comment]
Thanks, @rgommers !

**Comment by rgommers** [comment]
I'll close this, we completed this a while back; any remaining things that will pop up we can deal with in new issues. Thanks all!
